katello/katello-backup

```diff
@@ -63,59 +104,122 @@ def create_directories(directory)
   FileUtils.chmod_R 0770, directory
 end
 
-def backup_db_online
-  puts "Backing up postgres online schema... "
-  run_cmd("runuser - postgres -c 'pg_dumpall -g > #{@dir}/pg_globals.dump'")
-  run_cmd("runuser - postgres -c 'pg_dump -Fc foreman > #{@dir}/foreman.dump'")
-  run_cmd("runuser - postgres -c 'pg_dump -Fc candlepin > #{@dir}/candlepin.dump'")
-  puts "Done."
-
-  puts "Backing up mongo online schema... "
-  run_cmd('mongodump --host localhost --out mongo_dump')
-  puts "Done."
+def snapshot_backup
+  @mountdir = @options[:snapshot_mount] || "/var/snap"
+  @snapsize = @options[:snapshot_size] || "2G"
+  FileUtils.mkdir_p @mountdir
+  run_cmd("katello-service stop #{EXCLUDED}")
+  create_and_mount_snapshots
+  run_cmd("katello-service start #{EXCLUDED}")
+  backup_and_destroy_snapshots
 end
 
-def backup_pulp_online
-  puts "Backing up Pulp data... "
-  matching = false
- 
-  until matching
-    checksum1 = `find /var/lib/pulp -printf '%T@\n' | md5sum`
-    create_pulp_data_tar
-    checksum2 = `find /var/lib/pulp -printf '%T@\n' | md5sum`
-    matching = (checksum1 == checksum2)
+def create_and_mount_snapshots
+  @databases.each do |database|
+    puts "Creating and mounting #{database} snapshot"
+    lv_info = get_lv_info(database)
+    run_cmd("lvcreate -n#{database}-snap -L#{@snapsize} -s #{lv_info[0]}")
+    FileUtils.mkdir_p "#{File.join(@mountdir, database)}"
+    options = lv_info[1] == 'xfs' ? "-onouuid,ro" : "-oro"
+    run_cmd("mount #{get_snapshot_location(database)} #{File.join(@mountdir, database)} #{options}")
   end
-  puts "Done."
 end
 
-def backup_db_offline
-  puts "Backing up postgres db... "
-  run_cmd('tar --selinux --create --file=pgsql_data.tar --listed-incremental=.postgres.snar /var/lib/pgsql/data/')
-  puts "Done."
+def get_snapshot_location(database)
+  run_cmd("lvs --noheadings -o lv_path -S lv_name=#{database}-snap").strip
+end
 
-  puts "Backing up mongo db... "
-  run_cmd('tar --selinux --create --file=mongo_data.tar --listed-incremental=.mongo.snar --exclude=mongod.lock /var/lib/mongodb/')
-  puts "Done."
+def get_lv_info(database)
+  target = ""
+  case database
+  when 'pulp'
+    target = 'pulp'
+  when 'pgsql'
+    target = 'pgsql/data'
+  when 'mongod'
+    target = 'mongodb'
+  end
+  run_cmd("findmnt -n --target #{File.join('/var/lib', target)} -o SOURCE,FSTYPE").split
 end
 
-def backup_pulp_offline
-  puts "Backing up Pulp data... "
-  create_pulp_data_tar
-  puts "Done."
+def get_base_directory(database)
+  target = ""
+  case database
+  when 'pulp'
+    target = '0005_puppet_module_name_change.txt'
+  when 'pgsql'
+    target = 'postgresql.conf'
+  when 'mongod'
+    target = 'mongod.lock'
+  end
+  Find.find(File.join(@mountdir, database)) do |path|
```

Comment:
this loop will make the function return `nil` when when the target is not the last item in the list. I would suggest using something like this:

```ruby
  result = nil
  Find.find(File.join(@mountdir, database)) do |path|
    result = File.dirname(path) if File.basename(path) == target
  end
  result
end
```